src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java

```diff
@@ -49,37 +221,271 @@ public String ccloudOrganizationId() {
+          // Make sure we're not using both
+          if (sr != null && local != null && local.schemaRegistryUri != null) {
+            errors.add(
+                Error.create()
+                     .withDetail("Local config cannot be used with schema_registry configuration")
+                     .withSource("local_config.schema-registry-uri")
+            );
+          }
+        }
+        case CCLOUD -> {
+          localConfigNotAllowed(errors, newSpec);
+          kafkaClusterNotAllowed(errors, newSpec);
+          schemaRegistryNotAllowed(errors, newSpec);
+        }
+        case DIRECT -> {
+          var kafka = newSpec.kafkaClusterConfig();
+          if (kafka != null) {
+            if (kafka.bootstrapServers == null || kafka.bootstrapServers.isBlank()) {
+              isRequired(
+                  errors,
+                  "kafka_cluster.bootstrap_servers",
+                  "Kafka cluster bootstrap_servers"
+              );
+            }
+          }
+          var sr = newSpec.schemaRegistryConfig();
+          if (sr != null) {
+            if (sr.uri == null || sr.uri.isBlank()) {
+              isRequired(errors, "schema_registry.uri", "Schema Registry URI");
+            }
+          }
+          localConfigNotAllowed(errors, newSpec);
+          ccloudConfigNotAllowed(errors, newSpec);
+        }
+        case PLATFORM -> {
+        }
+        default -> {
+          errors.add(
+              Error.create()
+                   .withDetail("Unknown connection type: %s".formatted(newSpec.type()))
+                   .withSource("type")
+          );
+        }
+      }
     }
     return errors;
   }
+
+  void localConfigNotAllowed(List<Error> errors, ConnectionSpec newSpec) {
+    if (newSpec.localConfig != null) {
+      isNotAllowed(
+          errors,
+          LOCAL_CONFIG_FIELD_NAME,
+          "Local configuration",
+          "type is %s".formatted(newSpec.type)
+      );
+    }
+  }
+
+  void ccloudConfigNotAllowed(List<Error> errors, ConnectionSpec newSpec) {
+    if (newSpec.ccloudConfig != null) {
+      isNotAllowed(
+          errors,
+          CCLOUD_CONFIG_FIELD_NAME,
+          "CCloud configuration",
+          "type is %s".formatted(newSpec.type)
+      );
+    }
+  }
+
+  void kafkaClusterNotAllowed(List<Error> errors, ConnectionSpec newSpec) {
+    if (newSpec.kafkaClusterConfig != null) {
+      isNotAllowed(
+          errors,
+          KAFKA_CLUSTER_CONFIG_FIELD_NAME,
+          "Kafka cluster configuration",
+          "type is %s".formatted(newSpec.type)
+      );
+    }
+  }
+
+  void schemaRegistryNotAllowed(List<Error> errors, ConnectionSpec newSpec) {
+    if (newSpec.schemaRegistryConfig != null) {
+      isNotAllowed(
+          errors,
+          SCHEMA_REGISTRY_CONFIG_FIELD_NAME,
+          "Schema Registry configuration",
+          "type is %s".formatted(newSpec.type)
+      );
+    }
+  }
+
+  void isNotAllowed(List<Error> errors, String path, String what, String when) {
```

Comment:
I got used to using names that start with `is` for methods that return a boolean value. What about renaming this method to something like `addNotAllowedError` or similar?

Reply:
I renamed all of the methods to `check*(...)`, which I think is a decent combination of readable that avoids name patterns that are used for getters.